src/crypto/shelley/transactions/utxoTransactions.js

```diff
@@ -271,36 +272,126 @@ async function utxoToTxInput(utxo: RawUtxo): Input {
 async function addWitnesses(
   builderSetWitnesses: TransactionBuilderSetWitness,
   senderUtxos: Array<AddressedUtxo>,
-  wallet: Wallet.WalletObj,
+  signingKey: Bip32PrivateKey,
+  useLegacy: boolean,
 ): Promise<TransactionBuilderSetAuthData> {
   // get private keys
-  const rootKey = wallet.root_cached_key
   const privateKeys = await Promise.all(
     senderUtxos.map(
-      async (utxo): Promise<HdWallet.XPrv> => {
-        let key = rootKey
-        // this loops 3 times to generate a private key from root level up to the
-        // address level
-        for (let i = 0; i < Object.keys(utxo.addressing).length; i++) {
-          key = await HdWallet.derivePrivate(
-            key,
-            utxo.addressing[Object.keys(utxo.addressing)[i]],
-          )
-        }
-        return key
+      async (utxo): Promise<Bip32PrivateKey> => {
+        const chainKey = await signingKey.derive(utxo.addressing.change)
+        const addressKey = await chainKey.derive(utxo.addressing.index)
+        return addressKey
       },
     ),
   )
 
   const witnesses = await Witnesses.new()
   for (let i = 0; i < senderUtxos.length; i++) {
-    const witness = await Witness.for_utxo(
-      await Hash.from_hex(CONFIG.GENESISHASH),
-      // await txFinalizer.get_txid(),
-      await builderSetWitnesses.get_auth_data_for_witness(),
-      await v2SkKeyToV3Key(privateKeys[i]),
-    )
+    let witness
+    if (useLegacy) {
+      witness = await Witness.for_legacy_icarus_utxo(
+        await Hash.from_hex(CONFIG.GENESISHASH),
+        await builderSetWitnesses.get_auth_data_for_witness(),
+        privateKeys[i],
+      )
+    } else {
+      witness = await Witness.for_utxo(
+        await Hash.from_hex(CONFIG.GENESISHASH),
+        await builderSetWitnesses.get_auth_data_for_witness(),
+        await privateKeys[i].to_raw_key(),
+      )
+    }
     witnesses.add(witness)
   }
   return await builderSetWitnesses.set_witnesses(witnesses)
 }
+
+export const sendAllUnsignedTxFromUtxo = async (
+  receiver: string,
+  allUtxos: Array<RawUtxo>,
+): Promise<V3UnsignedTxData> => {
+  const totalBalance = allUtxos
```

Comment:
@v-almonacid probably we should throw if there isn’t any UTXO too. I don’t know if that could happen

Reply:
you mean if `allUtxos = []`? in that case `totalBalance` should be zero.